pkg/server/server_test.go

```diff
@@ -96,6 +97,124 @@ func BenchmarkOpenFGAServer(b *testing.B) {
+	mockDatastore.EXPECT().ReadUsersetTuples(gomock.Any(), storeID, gomock.Any()).AnyTimes().DoAndReturn(
+		func(_ context.Context, _ string, _ *openfgapb.TupleKey) (storage.TupleIterator, error) {
+			time.Sleep(50 * time.Millisecond)
+			return nil, errors.New("some error")
+		})
+	s := Server{
+		datastore: mockDatastore,
+		tracer:    telemetry.NewNoopTracer(),
+		meter:     telemetry.NewNoopMeter(),
+		transport: gateway.NewNoopTransport(),
+		logger:    logger.NewNoopLogger(),
+		config: &Config{
+			ResolveNodeLimit: 25,
+		},
+	}
+
+	checkResponse, err := s.Check(ctx, &openfgapb.CheckRequest{
+		StoreId:              storeID,
+		TupleKey:             tupleKey,
+		AuthorizationModelId: modelID,
+	})
+	require.NoError(t, err)
+	require.Equal(t, true, checkResponse.Allowed)
+}
+
+func TestShortestPathToSolutionWins(t *testing.T) {
+	t.SkipNow() //TODO remove
+	ctx := context.Background()
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	mockController := gomock.NewController(t)
+	defer mockController.Finish()
+
+	mockDatastore := mockstorage.NewMockOpenFGADatastore(mockController)
+	mockDatastore.EXPECT().ReadAuthorizationModel(gomock.Any(), storeID, modelID).AnyTimes().Return(&openfgapb.AuthorizationModel{
+		SchemaVersion: typesystem.SchemaVersion1_0,
+		TypeDefinitions: []*openfgapb.TypeDefinition{
+			{
+				Type: "repo",
+				Relations: map[string]*openfgapb.Userset{
+					"reader": typesystem.This(),
+				},
+			},
+		},
+	}, nil)
+
+	tupleKey := &openfgapb.TupleKey{
+		Object:   "repo:openfga",
+		Relation: "reader",
+		User:     "*",
+	}
+	tuple := &openfgapb.Tuple{Key: tupleKey}
+	// it could happen that one of the following two mocks won't be necessary because the goroutine will be short-circuited
+	mockDatastore.EXPECT().ReadUserTuple(gomock.Any(), storeID, gomock.Any()).AnyTimes().DoAndReturn(
+		func(_ context.Context, _ string, _ *openfgapb.TupleKey) (storage.TupleIterator, error) {
+			time.Sleep(500 * time.Millisecond)
+			return nil, storage.ErrNotFound
+		})
+	mockDatastore.EXPECT().ReadUsersetTuples(gomock.Any(), storeID, gomock.Any()).AnyTimes().DoAndReturn(
+		func(_ context.Context, _ string, _ *openfgapb.TupleKey) (storage.TupleIterator, error) {
+			time.Sleep(100 * time.Millisecond)
+			return storage.NewStaticTupleIterator([]*openfgapb.Tuple{tuple}), nil
+		})
+	s := Server{
+		datastore: mockDatastore,
+		tracer:    telemetry.NewNoopTracer(),
+		meter:     telemetry.NewNoopMeter(),
+		transport: gateway.NewNoopTransport(),
+		logger:    logger.NewNoopLogger(),
+		config: &Config{
+			ResolveNodeLimit: 25,
+		},
+	}
+
+	start := time.Now()
+	checkResponse, err := s.Check(ctx, &openfgapb.CheckRequest{
+		StoreId:              storeID,
+		TupleKey:             tupleKey,
+		AuthorizationModelId: modelID,
+	})
+	end := time.Since(start)
+	// we expect the Check call to be short-circuited after ReadUsersetTuples runs
+	require.Truef(t, end < 200*time.Millisecond, fmt.Sprintf("end was %s", end))
```

Comment:
Do you think this could be flaky? Once we have added resolution depth perhaps that will be a nicer thing to compare against?

Reply:
Yeah, it could be. And yeah, if resolution depth was in the response, we could check against that :(